dandi/download.py

```diff
@@ -501,3 +502,68 @@ def _download_file(
         os.utime(path, (time.time(), ensure_datetime(mtime).timestamp()))
 
     yield {"status": "done"}
+
+
+class DownloadDirectory:
+    def __init__(self, filepath, filehash):
+        #: The path to which to save the file after downloading
+        self.filepath = Path(filepath)
+        #: The expected hash of the downloaded data
+        self.filehash = filehash
+        #: The working directory in which downloaded data will be temporarily
+        #: stored
+        self.dirpath = append_ext(self.filepath, ".download")
+        #: The file in `dirpath` to which data will be written as it is
+        #: received
+        self.writefile = self.dirpath / self.filepath.name
+        #: A `fasteners.InterProcessLock` on `dirpath`
+        self.lock = None
+        #: An open filehandle to `writefile`
+        self.fp = None
+        #: How much of the data has been downloaded so far
+        self.offset = None
+
+    def __enter__(self):
+        from fasteners import InterProcessLock
+
+        self.dirpath.mkdir(parents=True, exist_ok=True)
+        self.lock = InterProcessLock(str(append_ext(self.writefile, ".lock")))
+        if not self.lock.acquire(blocking=False):
+            raise RuntimeError("Could not acquire lock on download")
+        chkpath = append_ext(self.writefile, ".checksum")
+        try:
+            checksum = chkpath.read_text(encoding="utf-8").strip()
+        except FileNotFoundError:
+            checksum = ""
+        if checksum == self.filehash:
+            # Pick up where we left off, writing to the end of the file
+            self.fp = self.writefile.open("ab")
+        else:
+            # Truncate the file (if it even exists)
+            self.fp = self.writefile.open("wb")
+            chkpath.write_text(str(self.filehash), encoding="utf-8")
+        self.offset = self.fp.tell()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if exc_type is None:
+            from shutil import rmtree
+
+            self.fp.close()
+            self.writefile.replace(self.filepath)
+            self.lock.release()
+            rmtree(self.dirpath, ignore_errors=True)
+        else:
+            self.fp.close()
+            self.lock.release()
```

Comment:
strangely `.close` is reported covered and `.lock.release()` -- not covered.  So `.fp.close()` raises exception somewhere in the tests?

Reply:
Aftering making the edits above, the pytest output shows that all of `__exit__` is covered.  I don't know why it wouldn't have been covered before.